[keba] Improve connection establishment and stability

MikeTheTux opened the pull request:
Signed-off-by: Michael Weger <weger.michael@gmx.net>

My second try of a pull request for the changes discussed in #6347 (comment)

I had some troubles with my first pull requst #10151 and therefore deleted it.

I still have the problem, that parts of this pull request have been developed by @Raketenschnitzel  and I do not know how to reach out to him in order to ask for his approval.
He might reply to this conversation, as he is mentioned ...

All other problems from the first PR should be resolved.


**[keba] Improve connection establishment and stability**

- Connection establishment no longer requires root permissions under Linux.
- Added support of additional P30 x-series stations 'E', 'G' & 'H'
- Extended add-on README
- introduced QuantityTypes (incompatible change)
- removed redundant pwmpilotcurrent (duplicate of maxpilotcurrent) (incompatible change)
- added maxpilotcurrentdutycyle

Binding for testing:
- [org.openhab.binding.keba-3.1.0-SNAPSHOT20210219.zip](https://github.com/openhab/openhab-addons/files/6012913/org.openhab.binding.keba-3.1.0-SNAPSHOT20210219.zip)
- [org.openhab.binding.keba-3.1.0-SNAPSHOT20210224.zip](https://github.com/openhab/openhab-addons/files/6033937/org.openhab.binding.keba-3.1.0-SNAPSHOT20210224.zip)

fwolter commented on bundles/org.openhab.binding.keba/src/main/java/org/openhab/binding/keba/internal/handler/KeContactHandler.java:
```diff
@@ -64,13 +66,14 @@
     private static final String CACHE_REPORT_2 = "REPORT_2";
     private static final String CACHE_REPORT_3 = "REPORT_3";
     private static final String CACHE_REPORT_100 = "REPORT_100";
+    public static final int SOCKET_TIME_OUT = 3000;
```
It's good practice to append the unit to the field name e.g. TIMEOUT_SEC.

MikeTheTux replied:
Ack

fwolter commented on bundles/org.openhab.binding.keba/src/main/java/org/openhab/binding/keba/internal/handler/KeContactHandler.java:
```diff
@@ -94,30 +97,49 @@ public KeContactHandler(Thing thing, KeContactTransceiver transceiver) {
-                } catch (Exception e) {
-                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.NONE,
-                            "An exception occurred while scheduling the polling job");
+        logger.debug("Initializing Keba binding");
```
Can this message be replaced by using the debugger or by increasing the framework's log level? See point 4 https://www.openhab.org/docs/developer/guidelines.html#f-logging

MikeTheTux replied:
fine. removed.

fwolter commented on bundles/org.openhab.binding.keba/src/main/java/org/openhab/binding/keba/internal/handler/KeContactHandler.java:
```diff
@@ -94,30 +97,49 @@ public KeContactHandler(Thing thing, KeContactTransceiver transceiver) {
+            if (isKebaReachable()) {
+                transceiver.registerHandler(this);
+
+                cache = new ExpiringCacheMap<>(
+                        Math.max((((BigDecimal) getConfig().get(POLLING_REFRESH_INTERVAL)).intValue()) - 5, 0) * 1000);
```
If the user forgets to configure this parameter in textual configuration, `get()` will return null.

MikeTheTux replied:
made the config parameter mandatory, as the entire binding implementation depends on it

fwolter replied:
See the other comment. It can be `null`, though.

MikeTheTux replied:
added a get'er `getRefreshInterval()` that handles `null.`
`getRefreshInterval()` and `getIPAddress()` are now always used when accessing the config

fwolter commented on bundles/org.openhab.binding.keba/src/main/java/org/openhab/binding/keba/internal/handler/KeContactHandler.java:
```diff
@@ -94,30 +97,49 @@ public KeContactHandler(Thing thing, KeContactTransceiver transceiver) {
+                    try {
+                        pollingJob = scheduler.scheduleWithFixedDelay(this::pollingRunnable, 0,
+                                ((BigDecimal) getConfig().get(POLLING_REFRESH_INTERVAL)).intValue(), TimeUnit.SECONDS);
+                    } catch (Exception e) {
```
What can throw an Exception here?

MikeTheTux replied:
scheduler throws ...
```
     * @throws RejectedExecutionException if the task cannot be
     *         scheduled for execution
     * @throws NullPointerException if command or unit is null
```

fwolter replied:
These are unchecked exceptions, which shouldn't be catched. The framework takes care of handling these appropriately.

MikeTheTux replied:
removed the exception handling for this code block

fwolter commented on bundles/org.openhab.binding.keba/src/main/java/org/openhab/binding/keba/internal/handler/KeContactHandler.java:
```diff
@@ -94,30 +97,49 @@ public KeContactHandler(Thing thing, KeContactTransceiver transceiver) {
-            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
-                    "IP address or port number not set");
+        } catch (Exception e) {
+            logger.error("Exception during Ping check and Http Backup check", e);
```
Bindings should only log to error if something severe happened, like the detection of a bug in your code. This could be debug. See this link for a description of the log levels: https://www.openhab.org/docs/developer/guidelines.html#f-logging

MikeTheTux replied:
Ack. Changed  to debug.

fwolter commented on bundles/org.openhab.binding.keba/src/main/java/org/openhab/binding/keba/internal/handler/KeContactHandler.java:
```diff
@@ -94,30 +97,49 @@ public KeContactHandler(Thing thing, KeContactTransceiver transceiver) {
-        } else {
-            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
-                    "IP address or port number not set");
+        } catch (Exception e) {
```
What type of exception do you expect here? Can you specify the concrete type?

MikeTheTux replied:
`IOException` from if `isKebaReachable()`
Concrete type specified

fwolter commented on bundles/org.openhab.binding.keba/src/main/java/org/openhab/binding/keba/internal/handler/KeContactHandler.java:
```diff
@@ -94,30 +97,49 @@ public KeContactHandler(Thing thing, KeContactTransceiver transceiver) {
+
+    private boolean isKebaReachable() throws IOException {
+        boolean isReachable = false;
+        SocketAddress sockAddr = new InetSocketAddress((String) getConfig().get(IP_ADDRESS), SOCKET_CHECK_PORT_NUMBER);
```
InetSocketAddress will return an `IllegalArgumentException` if the ip address is not set. Better check it beforehand whether it is null.

MikeTheTux replied:
Using getIPAddress() now. This function returns "" instead of null, if not configured (which is not possible, as the parameter is mandatory).
But even the null check is not sufficient. Even when != null it must not be a vaild IP address. The easiest way to figure that out is to try the init and catch the exception, isn't it?

fwolter replied:
If it is not a valid IP address, the string will be interpreted as a host name. If the host name can't be resovled, `Socket.connect()` will throw an `IOException`.

The IP address can't be `null` if the user uses the UI to configure the binding. But there will be no sanity checks if the user uses textual configuration. If the user doesn't specify the IP address, it will be null.

MikeTheTux replied:
Thanks for the hint with the textual config

MikeTheTux replied:
added a get'er `getRefreshInterval()` that handles `null.`
`getRefreshInterval()` and `getIPAddress()` are now always used when accessing the config

MikeTheTux commented:
@fwolter, thanks for the review!
please find my comments above.
I pushed my improvements into the branch.

it includes the following additional refactoring of the binding:
- introduced QuantityTypes (incompatible change)
- removed redundant pwmpilotcurrent (duplicate of maxpilotcurrent) (incompatible change)
- added maxpilotcurrentdutycyle

I hope it is OK to have this now in the same PR

MikeTheTux commented:
@fwolter, my latest commit should resolve all your recent comments

fwolter commented on bundles/org.openhab.binding.keba/src/main/resources/OH-INF/thing/kecontact.xml:
```diff
@@ -134,82 +134,82 @@
 		<label>Preset Current</label>
-		<description>Preset Current in mA</description>
-		<state pattern="%d mA" readOnly="false"></state>
+		<description>Preset Current in A</description>
+		<state pattern="%.3f A" readOnly="false"></state>
```
The unit shouldn't be mentioned in the description as the user can choose an arbitrary unit. You can use the placeholder in the pattern.
```suggestion
		<description>Preset Current</description>
		<state pattern="%.3f %unit%" readOnly="false" />
```

MikeTheTux replied:
done

fwolter commented on bundles/org.openhab.binding.keba/src/main/resources/OH-INF/thing/kecontact.xml:
```diff
@@ -21,7 +21,7 @@
 			<channel id="locked" typeId="locked"/>
 			<channel id="maxpilotcurrent" typeId="pilotcurrent"/>
-			<channel id="pwmpilotcurrent" typeId="pilotrange"/>
+			<channel id="maxpilotcurrentdutycyle" typeId="pilotcurrentdutycyle"/>
```
This change will break backward compatibility. Users, updating to OH 3.1 will need to adapt their config after updating. Are you sure about the change?

MikeTheTux replied:
Updating the config will anyhow be required, as the UoM support was introduced.

The old implementation having `pwmpilotcurrent` and `maxpilotcurrentdutycyle` didn't make sense, as both parameters have been updated with the same value, while `Max curr %` was not used at all:
```
                    case "Max curr": {
                        int state = entry.getValue().getAsInt();
                        maxPresetCurrent = state;
                        updateState(CHANNEL_PILOT_CURRENT, new DecimalType(state));
                        updateState(CHANNEL_PILOT_PWM, new DecimalType(state));
                        break;
                    }
```

The new implementation treats the duty cycle right. It is realized using the right type (`Number:Dimensionless` in %) as well as the name of the parameter.
```
                    case "Max curr": {
                        int state = entry.getValue().getAsInt();
                        maxPresetCurrent = state;
                        State newState = new QuantityType<ElectricCurrent>(state / 1000.0, Units.AMPERE);
                        updateState(CHANNEL_PILOT_CURRENT, newState);
                        break;
                    }
                    case "Max curr %": {
                        int state = entry.getValue().getAsInt();
                        State newState = new QuantityType<Dimensionless>(state / 10.0, Units.PERCENT);
                        updateState(CHANNEL_PILOT_PWM, newState);
                        break;
                    }
```

fwolter replied:
OK. Can you add a short message to https://github.com/openhab/openhab-distro/blob/7dcd280f91fcf9d014c0ca1529bb6ec770a27b4c/distributions/openhab/src/main/resources/bin/update.lst, explaining what has changed concerning backward compatibility?

MikeTheTux replied:
Is this the right file? There is no 3.1.0 nor 3.0.1 section yet. Is this the first breaking change for 3.1.0?

fwolter replied:
I was a bit puzzled, too. But it seems so.

MikeTheTux replied:
https://github.com/openhab/openhab-distro/pull/1263

MikeTheTux commented on bundles/org.openhab.binding.keba/src/main/java/org/openhab/binding/keba/internal/handler/KeContactHandler.java:
```diff
@@ -94,32 +103,45 @@ public KeContactHandler(Thing thing, KeContactTransceiver transceiver) {
-            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR,
-                    "IP address or port number not set");
+        } catch (IOException e) {
+            logger.debug("Exception during initialization of binding", e);
```
Implemented:
```
        } catch (IOException e) {
            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
                    "Exception during initialization of binding: " + e.toString());
        }
```
Tested:
`Thing 'keba:kecontact:f2c4d24b86' changed from INITIALIZING to OFFLINE (COMMUNICATION_ERROR): Exception during initialization of binding: java.net.SocketTimeoutException: connect timed out`

fwolter reviewed (approved):
LGTM
